agent/proxycfg/state.go

```diff
@@ -309,44 +339,202 @@ func (s *state) handleUpdate(u cache.UpdateEvent, snap *ConfigSnapshot) error {
+		}
+		svc := strings.TrimPrefix(u.CorrelationID, "upstream:"+serviceIDPrefix)
+		snap.UpstreamEndpoints[svc] = resp.Nodes
+
+	case strings.HasPrefix(u.CorrelationID, "upstream:"+preparedQueryIDPrefix):
+		resp, ok := u.Result.(*structs.PreparedQueryExecuteResponse)
+		if !ok {
+			return fmt.Errorf("invalid type for prepared query response: %T", u.Result)
+		}
+		pq := strings.TrimPrefix(u.CorrelationID, "upstream:")
+		snap.UpstreamEndpoints[pq] = resp.Nodes
+
 	default:
-		// Service discovery result, figure out which type
-		switch {
-		case strings.HasPrefix(u.CorrelationID, "upstream:"+serviceIDPrefix):
-			resp, ok := u.Result.(*structs.IndexedCheckServiceNodes)
-			if !ok {
-				return fmt.Errorf("invalid type for service response: %T", u.Result)
+		return errors.New("unknown correlation ID")
+	}
+	return nil
+}
+
+func removeColonPrefix(s string) (string, string, bool) {
+	idx := strings.Index(s, ":")
+	if idx == -1 {
+		return "", "", false
+	}
+	return s[0:idx], s[idx+1:], true
+}
+
+func (s *state) resetWatchesFromChain(
+	id string,
+	chain *structs.CompiledDiscoveryChain,
+	snap *ConfigSnapshot,
+) error {
+	if chain == nil {
+		return fmt.Errorf("not possible to arrive here with no discovery chain")
+	}
+
+	// Collect all sorts of catalog queries we'll have to run.
+	targets := make(map[structs.DiscoveryTarget]*structs.ServiceResolverConfigEntry)
+	addTarget := func(target structs.DiscoveryTarget) error {
+		resolver, ok := chain.Resolvers[target.Service]
+		if !ok {
+			return fmt.Errorf("missing resolver %q for target %s", target.Service, target)
+		}
+
+		targets[target] = resolver
+		return nil
+	}
+
+	// NOTE: We will NEVER see a missing chain, because we always request it with defaulting enabled.
+	for _, group := range chain.GroupResolverNodes {
+		groupResolver := group.GroupResolver
+
+		if err := addTarget(groupResolver.Target); err != nil {
+			return err
+		}
+		if groupResolver.Failover != nil {
+			for _, target := range groupResolver.Failover.Targets {
+				if err := addTarget(target); err != nil {
+					return err
+				}
 			}
-			svc := strings.TrimPrefix(u.CorrelationID, "upstream:"+serviceIDPrefix)
-			snap.UpstreamEndpoints[svc] = resp.Nodes
+		}
+	}
+
+	// Initialize relevant sub maps.
+	if _, ok := snap.WatchedUpstreams[id]; !ok {
+		snap.WatchedUpstreams[id] = make(map[structs.DiscoveryTarget]context.CancelFunc)
+	}
+	if _, ok := snap.WatchedUpstreamEndpoints[id]; !ok {
+		// TODO(rb): does this belong here?
+		snap.WatchedUpstreamEndpoints[id] = make(map[structs.DiscoveryTarget]structs.CheckServiceNodes)
+	}
 
-		case strings.HasPrefix(u.CorrelationID, "upstream:"+preparedQueryIDPrefix):
-			resp, ok := u.Result.(*structs.PreparedQueryExecuteResponse)
+	// We could invalidate this selectively based on a hash of the relevant
+	// resolver information, but for now just reset anything about this
+	// upstream when the chain changes in any way.
```

Comment:
I think Matt's PR does something closer to the hash thing. Not sure if that will be re-usable when we merge.

Reply:
I think the hash may have been excessive. Looking at my config entry replication code I think it sufficient to just check against the previous raft index of the primary dc

https://github.com/hashicorp/consul/blob/8d953f5840cf64e3db54c7cf0fbb92f12f50860a/agent/consul/config_replication.go#L34

You just need to ensure you don't check against the raft index of the local raft store.